core/services/bulletprooftxmanager/transmitchecker.go

```diff
@@ -0,0 +1,156 @@
+	) (bool, error)
+}
+
+func buildChecker(meta *EthTxMeta, client eth.Client) (TransmitChecker, error) {
+	if meta == nil {
+		return NoChecker{}, nil
+	}
+
+	switch meta.CheckerType {
+	case TransmitCheckerTypeSimulate:
+		return SimulateChecker{client}, nil
+	case TransmitCheckerTypeVRFV2:
+		coord, err := vrf_coordinator_v2.NewVRFCoordinatorV2(meta.VRFCoordinatorAddress, client)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to create coordinator at address %v", meta.VRFCoordinatorAddress)
+		}
+		return VRFV2Checker{coord.GetCommitment}, nil
+	case "":
+		return NoChecker{}, nil
+	default:
+		return nil, errors.Errorf("unrecognized checker type: %s", meta.CheckerType)
+	}
+}
+
+// NoChecker is a TransmitChecker that always determines a transaction should be submitted.
+type NoChecker struct{}
+
+func (_ NoChecker) ShouldTransmit(
+	_ context.Context,
+	_ logger.Logger,
+	_ EthTx,
+	_ EthTxAttempt,
+) (bool, error) {
+	return true, nil
+}
+
+type SimulateChecker struct {
+	Client eth.Client
+}
+
+func (s SimulateChecker) ShouldTransmit(
+	ctx context.Context,
+	l logger.Logger,
+	tx EthTx,
+	a EthTxAttempt,
+) (bool, error) {
+	simulationCtx, cancel := context.WithTimeout(ctx, simulationTimeout)
+	defer cancel()
+	// See: https://github.com/ethereum/go-ethereum/blob/acdf9238fb03d79c9b1c20c2fa476a7e6f4ac2ac/ethclient/gethclient/gethclient.go#L193
+	callArg := map[string]interface{}{
+		"from": tx.FromAddress,
+		"to":   &tx.ToAddress,
+		"gas":  hexutil.Uint64(a.ChainSpecificGasLimit),
+		// NOTE: Deliberately do not include gas prices. We never want to fatally error a
+		// transaction just because the wallet has insufficient eth.
+		// Relevant info regarding EIP1559 transactions: https://github.com/ethereum/go-ethereum/pull/23027
+		"gasPrice":             nil,
+		"maxFeePerGas":         nil,
+		"maxPriorityFeePerGas": nil,
+		"value":                (*hexutil.Big)(tx.Value.ToInt()),
+		"data":                 hexutil.Bytes(tx.EncodedPayload),
+	}
+	var b hexutil.Bytes
+	// always run simulation on "latest" block
+	err := s.Client.CallContext(simulationCtx, &b, "eth_call", callArg, eth.ToBlockNumArg(nil))
+	if err != nil {
+		if jErr := eth.ExtractRPCError(err); jErr != nil {
+			l.CriticalW("Transaction reverted during simulation",
+				"ethTxAttemptID", a.ID, "txHash", a.Hash, "err", err, "rpcErr", jErr.String(), "returnValue", b.String())
+			return false, errors.Errorf("transaction reverted during simulation: %s", jErr.String())
+		}
+		l.Warnw("Transaction simulation failed, will attempt to send anyway",
+			"ethTxAttemptID", a.ID, "txHash", a.Hash, "err", err, "returnValue", b.String())
+	} else {
+		l.Debugw("Transaction simulation succeeded",
+			"ethTxAttemptID", a.ID, "txHash", a.Hash, "returnValue", b.String())
+	}
+	return true, nil
+}
+
+type VRFV2Checker struct {
+	GetCommitment func(opts *bind.CallOpts, requestID *big.Int) ([32]byte, error)
+}
+
+func (v VRFV2Checker) ShouldTransmit(
+	_ context.Context,
```

Comment:
We _must_ pass in a context and allow timeouts here

Reply:
Done.